### PR TITLE
Use gc_alloc_obj for tapes containing Julia objects

### PR DESCRIPTION
**Pull request by vchuravy**
TODO:

- [x] We need to emit write_barriers on write of Julia objects to the Julia managed tape
- [x] Codegen for 1.6/1,7
- [x] Real runtime test, ideally with a nested tape.

### CODE REVIEW

File: src/compiler.jl [hunk 1]
```diff
@@ -2651,14 +2651,104 @@ function julia_error(cstr::Cstring, val::LLVM.API.LLVMValueRef, errtype::API.Err
...
+        return Any
+    else
+        return UInt # do we need something more precise?
+    end
+end
+
+to_tape_type(Type::LLVM.StructType) = Tuple{map(to_tape_type, LLVM.elements(Type))...}
+to_tape_type(Type::LLVM.ArrayType) = NTuple{Int(length(Type)), to_tape_type(eltype(Type))}
+function to_tape_type(Type::LLVM.IntegerType)
+    N = width(Type)
```
Comment (wsmoses) [inline src/compiler.jl:2683]: Add i1 support

File: src/compiler.jl [hunk 2]
```diff
@@ -2651,14 +2651,104 @@ function julia_error(cstr::Cstring, val::LLVM.API.LLVMValueRef, errtype::API.Err
...
+function to_tape_type(Type::LLVM.PointerType)
+    if 10 <= LLVM.addrspace(Type) <= 12
+        return Any
+    else
+        return UInt # do we need something more precise?
+    end
+end
+
+to_tape_type(Type::LLVM.StructType) = Tuple{map(to_tape_type, LLVM.elements(Type))...}
+to_tape_type(Type::LLVM.ArrayType) = NTuple{Int(length(Type)), to_tape_type(eltype(Type))}
```
Comment (wsmoses) [inline src/compiler.jl:2681]: Add vector type?

File: src/compiler.jl [hunk 3]
```diff
@@ -2680,11 +2770,16 @@ function julia_deallocator(B, Obj)
...
     mod = LLVM.parent(LLVM.parent(position(B)))
     ctx = context(mod)
-    ptr8 = LLVM.PointerType(LLVM.IntType(8; ctx))
+    
     T_void = LLVM.VoidType(ctx)
-    freeF = get_function!(mod, "free", LLVM.FunctionType(T_void, [ptr8]))
-    callf = call!(B, freeF, [pointercast!(B, Obj, ptr8)])
-    LLVM.API.LLVMAddCallSiteAttribute(callf, LLVM.API.LLVMAttributeIndex(1), EnumAttribute("nonnull"; ctx))
+    if any_jltypes(LLVM.llvmtype(Obj))
+        callf = LLVM.UndefValue(T_Void)
```
Comment (wsmoses) [inline src/compiler.jl:2776]: you should just return LLVM.Value(NULL) here

File: src/compiler.jl [hunk 4]
```diff
@@ -2651,14 +2651,104 @@ function julia_error(cstr::Cstring, val::LLVM.API.LLVMValueRef, errtype::API.Err
...
     Align = LLVM.Value(Align)
     LLVMType = LLVM.LLVMType(LLVMType)
-    mod = LLVM.parent(LLVM.parent(position(B)))
 
+    func = LLVM.parent(position(B))
+    mod = LLVM.parent(func)
     ctx = context(mod)
+
+    if any_jltypes(LLVMType)
+        @assert Count isa LLVM.ConstantInt
```
Comment (wsmoses) [inline src/compiler.jl:2717]: This count is not always 1....

File: src/compiler.jl (hunk 3 above)
Comment (vchuravy) [reply src/compiler.jl:2776]: I think I tried this previously and that didn't work

File: src/compiler.jl [hunk 5]
```diff
@@ -2651,14 +2651,104 @@ function julia_error(cstr::Cstring, val::LLVM.API.LLVMValueRef, errtype::API.Err
...
+    else
+        return UInt # do we need something more precise?
+    end
+end
+
+to_tape_type(Type::LLVM.StructType) = Tuple{map(to_tape_type, LLVM.elements(Type))...}
+to_tape_type(Type::LLVM.ArrayType) = NTuple{Int(length(Type)), to_tape_type(eltype(Type))}
+function to_tape_type(Type::LLVM.IntegerType)
+    N = width(Type)
+    if N == 8
```
Comment (vchuravy) [inline src/compiler.jl:2684]:
```suggestion
    if N == 1
        return Bool
    elseif N == 8
```

File: src/compiler.jl (hunk 3 above)
Comment (wsmoses) [reply src/compiler.jl:2776]: For freeing that should work, if not thats a bug on my end. For malloc, returning null is an error.

Comment (vchuravy) [reply src/compiler.jl:2776]: There is a Isa check on the free in Enzyme proper

Comment (wsmoses) [reply src/compiler.jl:2776]: Restore the return void and I'll change that do a != null and isa

File: src/compiler.jl [hunk 6]
```diff
@@ -3069,40 +3069,214 @@ function julia_error(cstr::Cstring, val::LLVM.API.LLVMValueRef, errtype::API.Err
...
+            EnzymeTapeT = LLVM.const_inttoptr(EnzymeTapeT, T_prjlvalue_UT)
+            EnzymeTapeT = LLVM.const_addrspacecast(EnzymeTapeT, T_prjlvalue)
+
+            TapeT = LLVM.ConstantInt(reinterpret(Int, Base.pointer_from_objref(TT)); ctx)
+            TapeT = LLVM.const_inttoptr(TapeT, T_prjlvalue_UT)
+            TapeT = LLVM.const_addrspacecast(TapeT, T_prjlvalue)
+
+            # call cc37 nonnull {}* bitcast ({}* ({}*, {}**, i32)* @jl_f_apply_type to {}* ({}*, {}*, {}*, {}*)*)({}* null, {}* inttoptr (i64 140150176657296 to {}*), {}* %4, {}* inttoptr (i64 140149987564368 to {}*))
+            tag = call!(B, f_apply_type, [LLVM.PointerNull(T_prjlvalue), EnzymeTapeT, boxed_count, TapeT])
+            LLVM.callconv!(tag, 37)
```
Comment (wsmoses) [inline src/compiler.jl:3229]: For 1.9+ this needs to be upgraded to a julia.call or julia.call2

**Comment by wsmoses** [comment]
Minimal remaining failure:
```
wmoses@beast:~/git/Enzyme.jl (vc/gc_alloc_obj) $ cat test/threads.jl 
using Enzyme
Enzyme.API.printall!(true)

    function foo(y)
        Threads.@threads for i in 1:3
            y[i] *= 2
        end
        nothing
    end

    x = [1.0, 2.0, 3.0]
    dx = [1.0, 1.0, 1.0]
    Enzyme.autodiff(Reverse, foo, Duplicated(x, dx))


    @show x

    @show dx

wmoses@beast:~/git/Enzyme.jl (vc/gc_alloc_obj) $ cat test/runtests.jl 
# HACK: work around Pkg.jl#2500
if VERSION < v"1.8-"
test_project = Base.active_project()
preferences_file = joinpath(dirname(@__DIR__), "LocalPreferences.toml")
test_preferences_file = joinpath(dirname(test_project), "LocalPreferences.toml")
if isfile(preferences_file) && !isfile(test_preferences_file)
    cp(preferences_file, test_preferences_file)
end
end

using Enzyme
using Test
using FiniteDifferences
using ForwardDiff
using Statistics
using LinearAlgebra

using Enzyme_jll
@info "Testing against" Enzyme_jll.libEnzyme

import LLVM_jll
using LLVM
# TODO: Add to LLVM_jll
function lit(; adjust_PATH=true, adjust_LIBPATH=true)
    lit_path = joinpath(LLVM_jll.artifact_dir, "tools", "lit", "lit.py")
    env = LLVM_jll.JLLWrappers.adjust_ENV!(
            copy(ENV),
            LLVM_jll.PATH[],
            LLVM_jll.LIBPATH[],
            adjust_PATH,
            adjust_LIBPATH,
        )
    return Cmd(Cmd([lit_path]); env)
end

@testset "Threads" begin
    cmd = `$(Base.julia_cmd()) --threads=1 --startup-file=no threads.jl`
   	@test success(pipeline(cmd, stderr=stderr, stdout=stdout))
end
```

```
$ ./julia-1.8.1/bin/julia --project -e "using Pkg; Pkg.test()"
signal (11): Segmentation fault
in expression starting at /home/wmoses/git/Enzyme.jl/test/threads.jl:18
getindex at ./array.jl:924 [inlined]
show_delim_array at ./show.jl:1211
show_delim_array at ./show.jl:1201 [inlined]
show_vector at ./arrayshow.jl:528
show_vector at ./arrayshow.jl:513 [inlined]
show at ./arrayshow.jl:484
unknown function (ip: 0x7f7af0761296)
_jl_invoke at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2367 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2549
#sprint#452 at ./strings/io.jl:114
sprint##kw at ./strings/io.jl:107 [inlined]
#repr#455 at ./strings/io.jl:282 [inlined]
repr at ./strings/io.jl:282
unknown function (ip: 0x7f7af075feb2)
_jl_invoke at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2367 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2549
jl_apply at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/julia.h:1838 [inlined]
do_call at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/interpreter.c:126
eval_value at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/interpreter.c:215
eval_stmt_value at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/interpreter.c:166 [inlined]
eval_body at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/interpreter.c:612
jl_interpret_toplevel_thunk at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/interpreter.c:750
jl_toplevel_eval_flex at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/toplevel.c:906
jl_toplevel_eval_flex at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/toplevel.c:850
ijl_toplevel_eval_in at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/toplevel.c:965
eval at ./boot.jl:368 [inlined]
include_string at ./loading.jl:1428
_jl_invoke at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2367 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2549
_include at ./loading.jl:1488
include at ./Base.jl:419
jfptr_include_32221.clone_1 at /home/wmoses/git/Enzyme.jl/julia-1.8.1/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2367 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2549
exec_options at ./client.jl:303
_start at ./client.jl:522
jfptr__start_61720.clone_1 at /home/wmoses/git/Enzyme.jl/julia-1.8.1/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2367 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/gf.c:2549
jl_apply at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/julia.h:1838 [inlined]
true_main at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/jlapi.c:575
jl_repl_entrypoint at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/src/jlapi.c:719
main at /cache/build/default-amdci5-0/julialang/julia-release-1-dot-8/cli/loader_exe.c:59
unknown function (ip: 0x7f7d23482d8f)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
Allocations: 34679443 (Pool: 34630854; Big: 48589); GC: 37
┌ Info: Testing against
└   Enzyme_jll.libEnzyme = "/home/wmoses/.julia/scratchspaces/7cc45869-7501-5eee-bdea-0790c847d4ef/build/Enzyme/libEnzyme-13.so"
```

**Review by wsmoses — APPROVED** [review]

File: src/compiler.jl (hunk 6 above)
Comment (wsmoses) [reply src/compiler.jl:3360]: @vchuravy I'm fine merging without this fix, but bumping here.

Comment (vchuravy) [reply src/compiler.jl:3360]: I would like to fix 1.9 support since that is the only thing that uses that branch

**Pull request merged**